### ISSUE TITLE
Introduce HTTPClient interface

### ISSUE THREAD

**Issue by mkabischev** [twitchtv/twirp#12]
It would be great to remove dependency to `http.Client` type. Instead of this interface can be used:
```go
// HTTPClient sends http.Requests and returns http.Responses or errors in case of failure.
type HTTPClient interface {
	Do(req *http.Request) (*http.Response, error)
}
```

It allows creating decorators for standard http.Client with some additional functionality like metrics, backoff retries, etc. `http.Client` implements this interface and such code is backward compatible with the current implementation. We use such approach and it's great.

What do you think? If you like this idea I can make PR.

**Comment by avidal** [comment]
fwiw, at a prior job we started by creating decorators that looked like an http.Client, but ran into a lot of problems, notably dependencies that want an actual `http.Client` instead of an interface.

What we ended up doing, which worked really well for us, was create `http.Transport` implementations instead. So we'd create a chain of transport decorators that were pluggable as the transport for any number of `http.Client` instances.

**Comment by spenczar** [comment]
Thanks for the suggestion, @mkabischev. I agree with @avidal that [`http.RoundTripper`](https://golang.org/pkg/net/http/#RoundTripper) wrappers are the way to go about this. The interface is very similar to what you suggest and is already in the standard library:

```
type RoundTripper interface {
  RoundTrip(*Request) (*Response, error)
}
```

Is there any advantage to wrapping at the `*http.Client` level instead that you're thinking of in particular?

**Comment by avidal** [comment]
Thanks for the clarification @spenczar; I got confused because our decorator repo package was called `transport`; but yes, they were implementations of `http.RoundTripper`.

**Comment by mkabischev** [comment]
Thanks for the response @spenczar & @avidal. I know about `http.RoundTripper` and internally we have the package that allows decorating both `RoundTripper` and our `Client` interface. `RoundTripper` decoration is required when some 3-rd party package depends on `*http.Client` and there is no other way to inject modified client. 

But in the documentation for `http.RoundTripper` there is the comment:

```
// RoundTrip should not attempt to interpret the response. In
// particular, RoundTrip must return err == nil if it obtained
// a response, regardless of the response's HTTP status code.
// A non-nil err should be reserved for failure to obtain a
// response. Similarly, RoundTrip should not attempt to
// handle higher-level protocol details such as redirects,
// authentication, or cookies.
//
// RoundTrip should not modify the request, except for
// consuming and closing the Request's Body.
```

So it means that injecting some information into the request (like tracking headers) is breaking of contract for this interface. The same is about checking response code and making retries. Also, RoundTripper will be invoked for every request in redirect chain.

Moving from `*http.Client` to interface doesn't break any existing code, allows to inject custom implementations and I think it's a little bit more idiomatic :)

**Comment by spenczar** [comment]
Okay, I understand this proposal better now - thank you @mkabischev.

I suppose we lose the ability to look at the `http.Client` struct's fields. We don't do that at all today, and I don't think we would do that with any of the fields it has currently, so it's not a big loss.

We'd want to define the accepted interface inline in the generated code, not in the `twirp` library, so that upgrading stays easy. Is there any problem with having the interface defined in each generated package? I don't _think_ there is.

Are there any other problems with accepting a `type HTTPClient interface { Do(*http.Request) (*http.Response, error) }` that people can think of?

**Comment by mkabischev** [comment]
Thanks. Defining interface in the generated package is better because then it will be placed next to the usage.

**Comment by rhysh** [comment]
> So it means that injecting some information into the request (like tracking headers) is breaking of contract for this interface.

The `http.RoundTripper` interface allows some flexibility here. Tracking headers themselves aren't a blocker: Yes the interface forbids modifying the `*http.Request`, but sending a _different_ `*http.Request` is fine. The `http.RoundTripper` can make a copy, add headers to that copy, and send the copy on the wire. The `WithContext` method can make the (shallow) copy for you.

Retries and redirects may still be an issue, depending on your infrastructure habits.

> Are there any other problems with accepting a `type HTTPClient interface { Do(*http.Request) (*http.Response, error) }` that people can think of?

- Code like `var myClient func(string, *http.Client) example.Haberdasher = example.NewHaberdasherProtobufClient` will stop compiling. Code like that sometimes shows up in tests, or in mocks.

- The package might already have a symbol with that name, maybe from a protobuf message or service called `HTTPClient`.

It's not 100% safe, but it's probably close.

**Comment by spenczar** [comment]
> It's not 100% safe, but it's probably close.
Man, I wish there were a way to know for sure!

I'm not worried about clashing with another type name - we can check that ourselves. I am a *little* worried about factory functions, but not terribly worried since it's at least just a change in the generated code.

Let's do it. This is ready for a PR if someone wants to be a new contributor :)

**Comment by spenczar** [comment]
This is now implemented in v5.1.0.
